crates/wasmparser/src/readers/core/types.rs

```diff
@@ -840,15 +1028,65 @@ pub struct TagType {
 }
 
 /// A reader for the type section of a WebAssembly module.
-pub type TypeSectionReader<'a> = SectionLimited<'a, Type>;
+pub type TypeSectionReader<'a> = SectionLimited<'a, SubType>;
 
-impl<'a> FromReader<'a> for Type {
+impl<'a> FromReader<'a> for StructuralType {
     fn from_reader(reader: &mut BinaryReader<'a>) -> Result<Self> {
+        read_structural_type(reader.read_u8()?, reader)
+    }
+}
+
+fn read_structural_type(
+    opcode: u8,
+    reader: &mut BinaryReader,
+) -> Result<StructuralType, BinaryReaderError> {
+    Ok(match opcode {
+        0x60 => StructuralType::Func(reader.read()?),
+        0x5e => StructuralType::Array(reader.read()?),
+        0x5f => StructuralType::Struct(reader.read()?),
+        x => return reader.invalid_leading_byte(x, "type"),
+    })
+}
+
+impl<'a> FromReader<'a> for SubType {
+    fn from_reader(reader: &mut BinaryReader<'a>) -> Result<Self> {
+        let pos = reader.original_position();
         Ok(match reader.read_u8()? {
-            0x60 => Type::Func(reader.read()?),
-            0x5e => Type::Array(reader.read()?),
-            0x5f => Type::Struct(reader.read()?),
-            x => return reader.invalid_leading_byte(x, "type"),
+            0x50 => {
+                let idx_iter = reader.read_iter(MAX_WASM_SUPERTYPES, "supertype idxs")?;
+                let idxs = idx_iter.collect::<Result<Vec<u32>>>()?;
+                if idxs.len() > 1 {
+                    return Err(BinaryReaderError::new(
+                        "multiple supertypes not supported",
+                        pos,
+                    ));
+                }
+                SubType {
+                    is_final: false,
+                    supertype_idx: idxs.first().copied(),
+                    structural_type: read_structural_type(reader.read_u8()?, reader)?,
+                }
+            }
+            0x4e => {
+                let idx_iter = reader.read_iter(MAX_WASM_SUPERTYPES, "supertype idxs")?;
+                let idxs = idx_iter.collect::<Result<Vec<u32>>>()?;
+                if idxs.len() > 1 {
+                    return Err(BinaryReaderError::new(
+                        "multiple supertypes not supported",
+                        pos,
+                    ));
+                }
+                SubType {
+                    is_final: true,
+                    supertype_idx: idxs.first().copied(),
+                    structural_type: read_structural_type(reader.read_u8()?, reader)?,
+                }
+            }
+            opcode => SubType {
+                is_final: false,
+                supertype_idx: None,
+                structural_type: read_structural_type(opcode, reader)?,
+            },
```

Comment:
There's a fair amount of duplication in this method, but could that perhaps be removed with something like:

```rust
let is_final = match reader.peek()? {
    0x50 | 0x4e => { /* advance reader */ }
    _ => None,
};
let supertype_idx = if is_final.is_some() {
    // read the supertype
} else {
    None 
};
let structural_type = /* preexisting structural type reading */;

SubType { is_final, supertype_idx, structural_type }
```

Reply:
Fixed.